TensorFlowSharp/nuget/build/net45/TensorFlowSharp.targets

```diff
@@ -0,0 +1,27 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+	<PropertyGroup>
+		<IsOSX Condition="Exists('/Library/Frameworks') and Exists ('/etc')">true</IsOSX>
+		<IsLinux Condition="Exists ('/proc') and Exists ('/etc/')">true</IsLinux>
+
+		<!-- if ShouldIncludeNativeTensorFlow == False then don't include the native libraries -->
+		<ShouldIncludeNativeTensorFlow Condition=" '$(ShouldIncludeNativeTensorFlow)' == '' ">True</ShouldIncludeNativeTensorFlow>
```

Comment:
I added an additional flag that could allow uses to excluded the bundled binaries, and use their own. This may be good if they want to use a GPU binary.

Reply:
Matt, perhaps you can help me here: how could I bundle both the CPU and GPU binaries, but have some user settable flag that controls which version to deploy with?

Reply:
One way would to actually not use `[DllImport]` but use delegates and `Marshal.GetDelegateForFunctionPointer`. This is what OpenGL.NET uses to load the various methods. The benefit of this is that it is switchable at runtime, but the downside is the code is a bit more complex and the native files must be named differently.

To do this at a packaging level, this is a bit harder as the .NET Core way of doing things is to just include native files - there is no way to have a "flagged" include. However, for normal, non-core projects, you can just add a simple element in the targets file.

Another way might be to split the packaging into three packages - one for the managed pinvokes, one for the GPU binaries and one for the CPU binaries. This way, the user installs TensorFlowSharp and say TensorFlowSharp.GPU. Then everything will work. This does add the extra overhead in which the user has to install two packages (and many will forget to do so), but this opens up a whole world of possibilities of other platforms being supported by third parties - as long as they have the same native API, they can use the normal TensorFlowSharp package (they just release a native binary package). In this case, third party libraries just depend on TensorFlowSharp, and only the final app has the TensorFlowSharp.CPU or TensorFlowSharp.GPU or event the Tizen.TensorFlowSharp.GPU...

Reply:
Ok, thanks for the explanation.   

Neither option sounds very appealing.   I am thinking that I might end up with something like having every P/Invoke listed twice, with different target platforms and switch at runtime based on a setting.

Reply:
This might only work if the CPU/GPU library names are named differently. The pinvoke usually requires that the native libraries live in the same folder, so they will have to be named differently.

Another downside to twice is that all work will have to be done twice - this may not be too bad if it is just a plain dll import, but is something to consider.